
# Lineage conversions table metadata

This table comes from https://github.com/jodyphelan/tbdb/blob/master/barcode.bed This is a barcode of all Fst = 1 SNPs generated by this paper - https://doi.org/10.1186/s13073-020-00817-3. All the SNP data has been stripped out and so just the unique lineages data has been left. 

The SNPs have been identified as belonging to an "MTBC lineage" - i.e. column 1, and at some point the other lineage columns have been joined via a lookup table. We're not sure what this is or where this has come from, but it seems to be right. It's unclear how these nomenclature/classification schemes map onto one another as I can't find anything equivalent to our table that makes it clear exactly *why*, for example, Lineage 3 = CAS = East-African-Indian. 

The supplementary material for **Cancino-Muñoz (2019)** (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6814786/) is the closest table that corresponds to our table (minus the RD numbers; they are referenced below in the **Faksri *et al* 2016** paper), and also references sources for the classification schemes. Although this paper states the sources are the "Nomenclature proposed by... " these seem unlikely to be the primary sources, however. For example, **Brudley, 2006** (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1468417/) contains a comprehensive list of spoligotype nomenclature, which pre-dates the **Demay 2012** paper below. I have not been able to track down the origins of the spoligotype nomenclature, nor the rational for how it maps to any other scheme (nor the reasons for why it's so complicated). 

Column headers and references for the **Cancino-Muñoz (2019)** table are as follows:

| MTBC lineage (1, 2) | LSP lineage (2) | Spoligotype family (4) |

1. Comas, I. et al. Human T cell epitopes of Mycobacterium tuberculosis are evolutionarily hyperconserved. Nat. Genet. 42, 498–503 (2010). 
2. Stucki, D. et al. Mycobacterium tuberculosis lineage 4 comprises globally distributed and geographically restricted sublineages. Nat. Genet. 48, 1535–1543 (2016).
3. Gagneux, S. et al. Variable host-pathogen compatibility in Mycobacterium tuberculosis. Proc. Natl. Acad. Sci. 103, 2869–2873 (2006).
4. Demay, C. et al. SITVITWEB – A publicly available international multimarker database for studying Mycobacterium tuberculosis genetic diversity and molecular epidemiology. Infect. Genet. Evol. 12, 755–766 (2012).

For a regions of difference-lineage conversion, see table 1 in **In silico region of difference (RD) analysis of Mycobacterium tuberculosis complex from sequence reads using RD-Analyzer, Faksri *et al* 2016**
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5093977/

This figure is from **Variable host–pathogen compatibility in Mycobacterium tuberculosis, Gagneux, DeRiemer *et al* 2006**  (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1413851/). Here the RD numbers and LSP (geographical) names are united in one figure. In the paper RDs seem to be synonymous with "LSP markers"; this makes sense since a "Region of Difference" is more-or-less the same thing as a "Large Sequence Polymorphism".
<img src=../img/Gagneux_DeRiemer_2006.png style="padding-right: 25px; padding-top: 25px;">
**Caption:** *The global population structure and geographical distribution of M. tuberculosis. (a) LSPs define a global phylogeny for M. tuberculosis. The names of the lineage-defining LSPs or regions of difference are shown in rectangles. The geographic regions associated with specific lineages are indicated. (b) The six main lineages of M. tuberculosis are geographically structured. Each dot corresponds to 1 of 80 countries represented in the global strain collection. The colors of the dots relate to the six main lineages defined in Fig. 1a and indicate the dominant lineage(s) in the respective countries.*


